---
title: "Figure 3 DGE Scatter & GSEA Bar plots for BMT Treg in Colon vs. non-Colon"
subtitle: "Figure 3 DGE Scatter comparing transplanted colon- vs. non-colon-homing Treg. GSEA Barplots for Genes up/down in colon-resident, CD62L- Treg for BMT Treg ranked according to colon vs. non-colon DGE comparison. Also, Supplementary Figure 3 subfigures based on the same subset of samples are generated."
author: "David Dittmar"
date: "`r Sys.Date()`"
bibliography: Rscripts.bib
nocite: '@MIRAGAIA2019493'
output:
  html_document:
    df_print: paged
    fig_caption: yes
    number_sections: no
    toc: yes
    toc_float: yes
    code_folding: hide 
---

***

```{r, include=FALSE}
# Load required libraries
library(edgeR)
library(AnnotationDbi)
library(org.Mm.eg.db) #for mouse
library(tibble)
library(dplyr)
library(ggplot2)
library(ggrepel)
library(ecoflux)
library(pryr)
#library(grid)
```

```{r setup, include=FALSE}
# Set global chunk options
knitr::opts_chunk$set(echo = TRUE, include = TRUE)
knitr::opts_knit$set(root.dir = normalizePath(".."))
# ...packages.bib file is also written here
knitr::write_bib(.packages(), "Figure3_TregDGEScatter_GSEA_cvnc.pc_BMT_packages.bib")
```

```{r}
# Load required counts, metadata and annotaion RDS files
counts.m<-readRDS("./RCTDIR/counts.m.rds")
metad.m<-readRDS("./METADIR/metad.m.rds")
stid<-readRDS("./METADIR/stid.rds")
```
* First, required data objects (counts, metadata, short transcript ID) and R packages are loaded. 
* This is the starting point for most analyses conducted on bulk RNA-seq data for the associated manuscript.

***

# Data Preparation
* For the analysis, the data and metadata have to be prepared and properly annotated.

## Annotation Data Frame
* An annotation data frame with 4 columns is constructed.
* The additional EntrezID column is required for possible downstream pathway analyses or annotation operations.
```{r}
genes.df<-as.data.frame(strsplit2(stid,"[$]"))
colnames(genes.df)<-c("EnsemblID","GeneSymbol","Length","GeneType")
genes.df$EnsemblID<-as.character(genes.df$EnsemblID)
genes.df$GeneSymbol<-as.character(genes.df$GeneSymbol)
genes.df$Length<-as.numeric(as.character(genes.df$Length))
genes.df$GeneType<-as.character(genes.df$GeneType)
genes.df$EntrezID<- mapIds(org.Mm.eg.db,
                    keys=genes.df$GeneSymbol,
                    column="ENTREZID",
                    keytype="SYMBOL",
                    multiVals="first")
rownames(genes.df)<-stid
head(genes.df, n=5L)
#See if the gene length values are still as they should, so this should yield an empty vector
which(strsplit2(stid,"[$]")[,3] %in% genes.df$Length==FALSE)
```

## Subsetting Index
* In this analysis, we focus on donor Treg re-isolated from recipient organs 7d after transfer (BMT w/o Tconv & BMT w Tconv).
* A subsetting index is generated accordingly.
```{r}
pc.ind<-(grepl("prc",metad.m$appl)|grepl("pro",metad.m$appl))&
  (grepl("S", metad.m$org)|grepl("C", metad.m$org)|grepl("L", metad.m$org))
print(paste(sum(pc.ind),"samples enter the analysis!"))
```

## Counts
* Here, the comprehensive counts data frame is reduced using the subsetting index.
```{r}
#Subsetting the respective counts object (counts.m)
counts.pc<-counts.m[,pc.ind]
paste(ncol(counts.pc),"samples are selected:")
colnames(counts.pc)
```

## Metadata
* The metadata has to be reduced accordingly. 
* The original order of samples is introduced for downstream re-ordering of the read count table.
* An additional group variable assigning colon-re-isolated samples to one group and spleen- or liver-migrated Treg to the other is introduced, as well.
* This enables the comparison of lymphoid vs. non-lymphoid tissue-homing Treg in the deployed BMT models, underlying this analysis.
```{r}
metad.pc<-metad.m[pc.ind,] %>% mutate(orig.ord = 1:sum(pc.ind)) %>% 
  arrange(grp) %>% arrange(org) %>% mutate(
    org.alt = case_when(
    org == "C" ~ "colon",
    org %in% c("L","S") ~ "spleen/liver"),
    organ = case_when(
      org=="C"~"colon",
      org=="L"~"liver",
      org=="S"~"spleen"))
counts.pc.ind<-metad.pc$orig.ord
head(metad.pc, n = 4L)
#the group variable for all later assignments of colors/pchs.
group.pc<-factor(as.character(metad.pc$grp))
group.pc
organ.pc <- factor(as.character(metad.pc$organ))
organ.pc
cvnc.pc <- factor(as.character(metad.pc$org.alt))
cvnc.pc
c.count<-sum(metad.pc$org %in% c("C"))
c.count
sl.count<-sum(metad.pc$org %in% c("L","S"))
sl.count
```

## Re-ordering Counts Data Frame Columns
* Here, read counts are re-ordered according to the metadata object.
```{r}
counts.pc.ord<-counts.pc[,counts.pc.ind]
head(counts.pc.ord, n=4L)
counts.pc<-counts.pc.ord
rm(counts.pc.ord)
```

## Consistency Check
* In order to avoid wrongful labelling, sample_IDs are used to confirm that the counts and metadata objects are consistent.
```{r}
if (sum(colnames(counts.pc)==metad.pc$sample_ID)==sum(pc.ind)) {
  print("Counts and metadata are properly subsetted and in accordance with each other. The analysis can be proceeded with!")
}else {
  print("Something is inconsistent, check the code above for errors during subsetting!")
}
```

## DGE List Object
* A DGEList object is generated, the main class of the edgeR package. 
* For full functionality, the counts.pct, group.avp.cvnc.pct and genes.df variables are used to do so.
```{r}
dgel.cvnc.pc<-DGEList(counts = counts.pc, group = organ.pc, 
                      genes = genes.df)
head(dgel.cvnc.pc$samples, n=4L)
head(dgel.cvnc.pc$genes, n=4L)
```

## Exclusion of Rare Transcripts
* The "filterByExpr()" function is used to determine which genes have sufficiently large counts to be retained in a statistical analysis. 
* Only genes fulfilling that criterion are kept and also normalization factors are calculated.
```{r}
keep.cvnc.pc <- filterByExpr(dgel.cvnc.pc)
dgel.cvnc.pc <- dgel.cvnc.pc[keep.cvnc.pc, , keep.lib.sizes=FALSE]
dgel.cvnc.pc <- calcNormFactors(dgel.cvnc.pc)
print(paste(table(keep.cvnc.pc)[1],"genes are discarded while",table(keep.cvnc.pc)[2], "genes are retained in the analysis!"))
```

***

# Differential Gene Expression (DGE) analysis
* Differential Gene Expression (DGE) is performed in the following part.

## Design for DGE
* Hereafter, a design is created that contains the contrast of interest.
* Blocking factors that can be adjusted are included in the model, as well.
```{r}
# Contrast of interest
cvnc.pc
# Blocking Factor
pcr2<-factor(metad.pc$PCR2.Cy)
treat<-factor(metad.pc$treat)
appl<-factor(metad.pc$appl)
# Design
cvnc.pc.design.dge<-model.matrix(~0+cvnc.pc+appl+treat+pcr2)
colnames(cvnc.pc.design.dge)<-c("colon", "liver_spleen", "GvHD", "poly", 
                                "twelve", "thirteen", "fourteen")
as_tibble(cvnc.pc.design.dge)
```

## Dispersion Esitmation
* Hereafter, the dispersion is estimated.
```{r}
dgel.cvnc.pc<-estimateDisp(dgel.cvnc.pc,design=cvnc.pc.design.dge,robust=TRUE)
print(paste0("The common dispersion amounts to ",
             round(dgel.cvnc.pc$common.dispersion, digits = 8)))
```

## GLM Fitting
* Fitting gene wise glm (generalized linear model) for the constructed design.
```{r}
cvnc.pc.fit<-glmQLFit(dgel.cvnc.pc,cvnc.pc.design.dge)
```

## DGE for Contrast of Interest (COI)
* The interest in this analysis lies in the differences between Treg re-isolated from colon versus spleen&liver. 
* Thus, our contrast of interest is "colon - liver_spleen" for the DGE analysis. 
* We can have a quick look at how many genes are DE with a FC > 1.5 and FDR < 0.05.
```{r}
# Contrast of Interest
cvnc.pc.con<-makeContrasts(colon - liver_spleen, levels=cvnc.pc.design.dge) 
# DGE testing against fc = 1.5.
cvnc.pc.tr<-glmTreat(cvnc.pc.fit, contrast=cvnc.pc.con, lfc=log2(1.5))
cvnc.pc.tr.qstat<-topTags(cvnc.pc.tr,n = Inf)
cvnc.pc.tr.qdt<-decideTestsDGE(cvnc.pc.tr)
summary(cvnc.pc.tr.qdt)
```

### Exporting DGE Statistics
* The data frame contained in the glmTreat function results is stored for direct access.
```{r}
write.table(x = cvnc.pc.tr.qstat$table,
            file = "./ANALDIR/Figure3b_TregDGE_cvnc_pro.prc_FC1.5.txt",
            quote=FALSE, 
            sep = "\t",
            col.names = TRUE,
            row.names = FALSE)
```

***

# DGE Visualization
* Hereafter, DEGs are visualized *via* a scatter plot.
* In Addition, Genes up/down in colon-resident Cd62L- Treg will be ranked according to log2(FC) of the respective comparison in BMT Treg (w & w/o Tconv).

## Data Preparation
* The data is filtered based on logCPM and logRPKM thresholds.
* Necessary data transformations for the scatter plot are carried out as well: columns with individual logCPMs (calculated from logCPM and logFC)
* The total number of genes, up and down regulated genes is printed.
```{r}
# Extraction of data frame from glmTreat result
cvnc.pc.tr.qstat.scat <- cvnc.pc.tr.qstat$table %>% #is a data frame
  # Introduction of logRPKM column
  mutate(logRPKM=logCPM-log2(Length * 0.001))
# Commonly used filter criteria
cvnc.pc.tr.qstat.scat.filt <- filter(cvnc.pc.tr.qstat.scat, 
                                      logCPM > 1 & logRPKM > 1) %>%
  # Transformation for scatter plot --> individual log CPMs
  mutate(logCPMnonColon = logCPM - logFC/2, # non-colon
         logCPMcolon = logCPM + logFC/2) # colon
nrow(cvnc.pc.tr.qstat.scat.filt)
# Separation of significantly up and down regulated genes
cvnc.pc.tr.qstat.scat.filt.UP<-filter(cvnc.pc.tr.qstat.scat.filt, logFC >= 0 &
                                        FDR < 0.05)
cvnc.pc.tr.qstat.scat.filt.DN<-filter(cvnc.pc.tr.qstat.scat.filt, logFC <= 0 &
                                        FDR < 0.05)
nrow(cvnc.pc.tr.qstat.scat.filt.UP)
nrow(cvnc.pc.tr.qstat.scat.filt.DN)
```

## Exporting UP and DN regulated Genes
* Significantly up and down regulated genes are exported in the following.
```{r}
#extract filtered gene lists
write.table(rownames(cvnc.pc.tr.qstat.scat.filt.UP),
            file = "./ANALDIR/Figure3b_TregDGE_cvnc_pro.prc.filt.sig.UP.txt",
            col.names = "GeneID",
            quote=FALSE, 
            row.names = FALSE)
write.table(rownames(cvnc.pc.tr.qstat.scat.filt.DN),
            file = "./ANALDIR/Figure3b_TregDGE_cvnc_pro.prc.filt.sig.DN.txt",
            col.names = "GeneID",
            quote=FALSE, 
            row.names = FALSE)
```

## Scatterplot

### Definition of Genes of Interest (GOI)
* Hereafter, GOI are taken from Miragaia et al. 2019 (Figure 2 B) presenting Treg subpopulations of the steady-state.
* The question addressed here is whether our data is in line with their findings. 
* [Here you can have a look at the figure](https://pubmed.ncbi.nlm.nih.gov/30737144/#&gid=article-figures&pid=figure-2-uid-2)
```{r}
# non-lymphoid tissue
tm_fig2_panel_NLT<-rev(c("Ccr1","Ccr2","Gzmb","Il10","Il1rl1","Kdm6b","Gata3",
                      "Areg","S100a4","Itgae","Rora","Klrg1","Lgals1","Fgl2",
                      "Cmtm7","Cd44","Icos","Ctla4","Pim1","Ikzf2","Pdcd1",
                      "Cd83","Relb","Tnfrsf9","Tnfrsf4"))
# lymphoid tissue
tm_fig2_panel_LT<-rev(c("Stat1","Ccr7","S1pr1","Sell","Bcl2","Tcf7"))
```

### Color-coded geom_text_repel data frames
* First, to be able to label data points for all categories with a single call to geom_text_repel, additional data frames are generated.
```{r}
cvnc.pc.qstat.scat.filt.ann <- cvnc.pc.tr.qstat.scat.filt %>%
  mutate(
    ann.col = case_when(
      GeneSymbol %in% tm_fig2_panel_NLT ~ "brown4",
      GeneSymbol %in% tm_fig2_panel_LT ~ "navy"),
    goi.cat = case_when(
      logFC <= 0 & FDR < 0.05 ~ paste0("down in colon (",as.character(nrow(cvnc.pc.tr.qstat.scat.filt.DN)),")"),
      logFC >= 0 & FDR < 0.05 ~ paste0("up in colon (",as.character(nrow(cvnc.pc.tr.qstat.scat.filt.UP)),")"),
      TRUE ~ "non DE")
  )
cvnc.pc.qstat.scat.filt.ann.label <- filter(cvnc.pc.qstat.scat.filt.ann, !is.na(ann.col))
```
### Scatterplot with GOI highlighted (Figure 3 b)
* Hereafter, the scatter plot is generated and plotted.
```{r, fig.height=2.6, fig.width=4.2, warning=FALSE}
cvnc.pc.scat.colon.tm <- ggplot(cvnc.pc.tr.qstat.scat.filt) + 
  aes(x=logCPMnonColon, y=logCPMcolon) +
  scale_x_continuous(limits = c(-2,14),
                     breaks = c(0,3,6,9,12),
                     expand = c(0,0)
                     )+
  scale_y_continuous(limits = c(-2,14),
                     breaks = c(0,3,6,9,12),
                     expand = c(0,0)
                     )+
  geom_point(colour=rgb(50,50,50,30,maxColorValue = 255), shape=16, size = .375)+
  geom_point(data=filter(cvnc.pc.qstat.scat.filt.ann, goi.cat != "non DE"),
             aes(x=logCPMnonColon, y=logCPMcolon, color=goi.cat), 
             shape=16, size = .5, alpha=.75) +
  scale_color_manual(name = "Significant DEGs",
                     values = c("slateblue","sienna"),
                     guide = guide_legend(override.aes = list(size = 3)))+
  geom_text_repel(data = cvnc.pc.qstat.scat.filt.ann.label,
                  aes(label=GeneSymbol, color = ann.col),
                  seed = 42,
                  direction = "both",
                  force = 500,
                  force_pull = 0,
                  size=1.66,
                  min.segment.length=0.0,
                  segment.size=0.15,
                  segment.color = cvnc.pc.qstat.scat.filt.ann.label$ann.col,
                  color = cvnc.pc.qstat.scat.filt.ann.label$ann.col,
                  max.overlaps = 25,
                  point.padding = 0.0,
                  fontface="italic")+
  ggtitle(label = "NLT & LT signature in BMT Treg") +
  #xlab(bquote(log[2]~CPM *"("*liver*"&"*spleen*")"))+
  #ylab(bquote(log[2]~CPM *"("*colon*")"))+
  xlab(bquote(log[2]~CPM ~liver*"&"*spleen*" (n=47)"))+
  ylab(bquote(log[2]~CPM ~colon*" (n=24)"))+
  theme(plot.title = element_text(size = 10, face = "bold"),
        plot.subtitle = element_text(size = 6, face = "bold"),
        axis.title = element_text(size = 8),
        axis.text.x = element_text(size = 6, colour = "black"),
        axis.text.y =element_text(size = 6, colour = "black"),
        axis.ticks.length.x = unit(0.1,"cm"),
        axis.ticks.length.y = unit(0.1,"cm"),
        axis.ticks = element_line(size=.236),
        plot.background = element_rect(fill="white",colour = "white"),
        panel.background = element_rect(fill = NA,color = "black", size=.236),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        legend.key = element_blank(),
        aspect.ratio = 1)
cvnc.pc.scat.colon.tm
```

* The scatterplot is also exported into the figure directory.

```{r, warning=FALSE}
pdf(file="./FIGDIR/Figure3b_TregDGEScatter_cvnc.pc_tmGenes.pdf", 
    height=2.6, width=4.2)
cvnc.pc.scat.colon.tm
dev.off()
```

***

# GSEA with baseline colon signatures
* Please note, you should have run "Figure3_TregDGEScatter_GSEA_Heatmap_Metascape_cvnc.b.Cd62Lneg.Rmd" before proceeding.
* The question addressed here is, whether differential genes in organ-resident Treg also tend to be differentially expressed in transplanted Treg.

## Reading in DGE from Baseline Colon Treg
* Data from analogous baseline Treg analysis is read in, hereafter. 
```{r}
bc.up <- read.delim(
  file = "./ANALDIR/SupplementaryFigure3i_cluster4.colon.up.GeneIDs.txt")
bc.dn <- read.delim(
  file = "./ANALDIR/SupplementaryFigure3i_cluster1.colon.down.GeneIDs.txt")
```

* First, indices have to be generated for the underlying generalized linear model of this analysis.
* Using the fry() function, a rotation gene set testing for linear models is carried out.

```{r}
ind1 <- rownames(cvnc.pc.fit) %in% bc.up[[1]]
ind2 <- rownames(cvnc.pc.fit) %in% bc.dn[[1]]
ind <- list(ind1,ind2)
names(ind) <- c("up in colon","down in colon")
gse.cvnc.pc.treat.lfc0.6 <- fry(dgel.cvnc.pc, index=ind, 
                                design=cvnc.pc.design.dge ,contrast=cvnc.pc.con)
gse.cvnc.pc.treat.lfc0.6
```

## GSEA Barcode plot (Figure 3 c) 
* Herafter, the actual barcode plot is generated and visualized.
* The adjusted p value for the enrichments visualized is extracted from the fry results.
```{r, fig.height=5, fig.width=5}
cvnc.pc.bc.b.cvnc %<a-% {
  par(mar=c(4.1,2.1,0.1,0.1))
  barcodeplot(cvnc.pc.tr$table$logFC,
              index=ind1,
              index2=ind2,
              labels=c("down in colon","up in colon"),
              xlab = bquote(log[2]*"FC"),
              col.bars = c("peru","lightslateblue"))
par(new=TRUE)
plot.new( )
plot.window( xlim=c(-5,5), ylim=c(-5,5) )
text(-5,5,"Brown bars:\nGenes up in resident colon Treg", adj = c(0,.5))
text(c(-5,-3.75),c(3.4,3.5),
     c(expression(italic(P)["adj."]*"<"),
       scientific_10x(
         plyr::round_any(
           x = gse.cvnc.pc.treat.lfc0.6["up in colon",]$FDR,
           accuracy = 10^(ceiling(
             log10(gse.cvnc.pc.treat.lfc0.6["up in colon",]$FDR))-2),
           f = ceiling),
         digits = 1)), adj = c(0,.5))
text(5.0,-3.8,"Steelblue bars:\nGenes down resident colon Treg", adj = c(1,.5))
text(c(3.0, 5.0),c(-4.7,-4.6),
     c(expression(italic(P)["adj."]*"<"),
       scientific_10x(
         plyr::round_any(
           x = gse.cvnc.pc.treat.lfc0.6["down in colon",]$FDR,
           accuracy = 10^(ceiling(
             log10(gse.cvnc.pc.treat.lfc0.6["down in colon",]$FDR))-2),
           f = ceiling),
         digits = 1)),adj = c(1,.5))
}
cvnc.pc.bc.b.cvnc
```

* There is a large overlap between resident and BMT samples: Hence transplanted Treg acquire organ phenotype rapidly.
* The barcode plot is exported, hereafter.

```{r}
pdf(file="./FIGDIR/Figure3c_cvnc.pc.Barcodeplot_residentColonSignatures.pdf", height=5, width=5)
cvnc.pc.bc.b.cvnc
dev.off()
```

***

# Additional DGE Analysis on the same Subset of samples
* On the same subset of samples (all Treg re-isolated from tissues following transplant), a DGE analysis is performed.
* The comparison of "allo" vs. "poly" Treg is addressed here.

## Metadata
* For better readability, a new group variable is introduced
```{r}
metad.pc<-metad.pc %>% mutate(
  treat.alt=case_when(
    treat == "a" ~ "allo_BMT",
    treat == "p" ~ "poly_BMT"
  )
)
#the group variable for all later assignments of colors/pchs.
avp.pc<-factor(as.character(metad.pc$treat.alt))
avp.pc
allo.count<-sum(avp.pc == "allo_BMT")
allo.count
poly.count<-sum(avp.pc == "poly_BMT")
poly.count
```

## DGE List Object
* For the purpose of this analysis we require an additional DGEList object.
* A fitting group variable is defined to do so.
```{r}
dgel.avp.pc<-DGEList(counts = counts.pc, group = group.pc, genes = genes.df)
head(dgel.avp.pc$samples, n=4L)
head(dgel.avp.pc$genes, n=4L)
```

## Exclude Rare Transcripts
* The "filterByExpr()" function is used to determine which genes have sufficiently large counts to be retained in a statistical analysis. 
* Only genes fulfilling that criterion are kept and also normalization factors are calculated.
```{r}
keep.avp.pc <- filterByExpr(dgel.avp.pc)
dgel.avp.pc <- dgel.avp.pc[keep.avp.pc, , keep.lib.sizes=FALSE]
dgel.avp.pc <- calcNormFactors(dgel.avp.pc)
print(paste(table(keep.avp.pc)[[2]],"genes remain in the analysis after filtering!"))
```

***

# Differential Gene Expression (DGE) analysis
* Differential Gene Expression (DGE) is performed in the following part.

## Design for DGE
In the design for DGE it is critical to include the following factors:

* avp.pc: allo/ poly, which is our contrast of interest
* appl: prophylaxis/control as blocking factor
* org: colon/liver/spleen as blocking factors
* pcr2: this is the only applicable technical variable for batch-correction and should also be included as blocking factor

```{r}
# contrast of interest
avp.pc
# blocking factors
appl
organ.pc
pcr2
avp.pc.design.dge<-model.matrix(~0+avp.pc+appl+organ.pc+pcr2)
colnames(avp.pc.design.dge)<- c(levels(avp.pc),"prophylaxis","liver","spleen","twelve","thirteen","fourteen")
as_tibble(avp.pc.design.dge)
```

## Dispersion Estimation
* Hereafter, dispersion is estimated.
```{r}
dgel.avp.pc<-estimateDisp(dgel.avp.pc,design=avp.pc.design.dge,robust=TRUE)
dgel.avp.pc$common.dispersion #0.1035074
```

## GLM Fitting
* Fitting gene wise glm (generalized linear model) for the constructed design.
```{r}
avp.pc.fit<-glmQLFit(dgel.avp.pc,avp.pc.design.dge)
```

## DGE for Contrast of Interest (COI)
* The difference between allo and poly Treg afer transplantation and re-isolation is of interest, hereafter.
* Thus, our contrast of interest is "allo - poly" for the DGE analysis. 
* We can have a quick look at how many genes are DE with a FC > 1.5 and FDR < 0.05.
```{r}
#Contrast of intrest
avp.pc.con<-makeContrasts(allo_BMT - poly_BMT, levels=avp.pc.design.dge)
# DGE testing against fc = 1.5
avp.pc.tr <- glmTreat(avp.pc.fit, contrast=avp.pc.con, lfc=log2(1.5))
avp.pc.tr.qstat<-topTags(avp.pc.tr,n = Inf)
avp.pc.tr.qdt<-decideTestsDGE(avp.pc.tr)
summary(avp.pc.tr.qdt)
```

### Exporting DGE Statistics
* The data frame contained in the glmTreat function results is stored for direct access.
```{r}
write.table(x = avp.pc.tr.qstat$table,
            file = "./ANALDIR/SupplementaryFigure3c_TregDGE_avp_pro.prc_FC1.5.txt",
            quote=FALSE, 
            sep = "\t",
            col.names = TRUE,
            row.names = FALSE)
```

***

# DGE Visualization
* Hereafter, DEGs are visualized *via* a scatter plot.

## Data Preparation
* The data is filtered based on logCPM and logRPKM thresholds.
* Necessary data transformations for the scatter plot are carried out as well: columns with individual logCPMs (calculated from logCPM and logFC)
```{r}
# Extraction of data frame from glmTreat result
avp.pc.tr.qstat.scat <- avp.pc.tr.qstat$table %>% #is a data frame
  # Introduction of logRPKM column
  mutate(logRPKM=logCPM-log2(Length * 0.001))
# Commonly used filter criteria
avp.pc.tr.qstat.scat.filt <- filter(avp.pc.tr.qstat.scat, 
                                      logCPM > 1 & logRPKM > 1) %>%
  # Transformation for scatter plot --> individual log CPMs
  mutate(logCPMpoly = logCPM - logFC/2, # non-colon
         logCPMallo = logCPM + logFC/2) # colon
nrow(avp.pc.tr.qstat.scat.filt)
# Separation of significantly up and down regulated genes
avp.pc.tr.qstat.scat.filt.UP<-filter(avp.pc.tr.qstat.scat.filt, logFC >= 0 &
                                        FDR < 0.05)
avp.pc.tr.qstat.scat.filt.DN<-filter(avp.pc.tr.qstat.scat.filt, logFC <= 0 &
                                        FDR < 0.05)
nrow(avp.pc.tr.qstat.scat.filt.UP)
nrow(avp.pc.tr.qstat.scat.filt.DN)
```

## Export UP and DN regulated Genes
* Significantly up and down regulated genes are exported in the following.
```{r}
#extract filtered gene lists
write.table(rownames(avp.pc.tr.qstat.scat.filt.UP),
            file = "./ANALDIR/SupplementaryFigure3c_TregDGE_avp_pro.prc.filt.sig.UP.txt",
            col.names = "GeneID",
            quote=FALSE, 
            row.names = FALSE)
write.table(rownames(avp.pc.tr.qstat.scat.filt.DN),
            file = "./ANALDIR/SupplementaryFigure3c_TregDGE_avp_pro.prc.filt.sig.DN.txt",
            col.names = "GeneID",
            quote=FALSE, 
            row.names = FALSE)
```

## Scatterplot

### Definition of Genes of GOI
* Genes of interest defining Treg in general are defined in this step.
* Approach: Intersection of "Treg-meta-signature" and "Immnav-Treg-signature" by Aubert et al. 2020, extended with genes Itgb7 and Sell.
```{r}
# Reading in the intersection of "Treg-meta-signature" and "Immnav-Treg-signature" published by Aubert et al. 2020
aubert.treg.intersect<-readRDS(file ="./PUBDATDIR/aubert.treg.intersect.rds")
# Extension of the GOI by Itgb7 and Sell
aubert.treg.intersect<-c(aubert.treg.intersect,"Itgb7","Sell")
# Further added examples of differentially regulated genes
aubert.treg.intersect<-c(aubert.treg.intersect,"Lag3","Nrn1","Nrgn","Rbpj","Ccdc184")
aubert.treg.intersect
```

### Color-coded geom_text_repel data frames
* First we prepare some labels to be able to label data points for all categories with a single call to geom_text_repel.
```{r}
avp.pc.qstat.scat.filt.ann<-avp.pc.tr.qstat.scat.filt %>%
  mutate(
    ann.col = case_when(
      GeneSymbol %in% intersect(avp.pc.tr.qstat.scat.filt.UP$GeneSymbol,aubert.treg.intersect) ~ "orange",
      GeneSymbol %in% intersect(avp.pc.tr.qstat.scat.filt.DN$GeneSymbol,aubert.treg.intersect) ~ "steelblue",
      GeneSymbol %in% aubert.treg.intersect ~ "dimgrey"),
    goi.cat = case_when(
      logFC >= 0 & FDR < 0.05 ~ paste0("up in allo (",as.character(nrow(avp.pc.tr.qstat.scat.filt.UP)),")"),
      logFC <= 0 & FDR < 0.05 ~ paste0("down in allo (",as.character(nrow(avp.pc.tr.qstat.scat.filt.DN)),")"),
      TRUE ~ "non DE"),
    goi.cat.col = case_when(
      logFC >= 0 & FDR < 0.05 ~ "orange",
      logFC <= 0 & FDR < 0.05 ~ "steelblue")
  )
# reduce df to labels to be printed in plot
avp.pc.qstat.scat.filt.ann.label <- filter(avp.pc.qstat.scat.filt.ann, !is.na(ann.col))
nrow(avp.pc.qstat.scat.filt.ann.label)
# number of allo and poly samples
nrow(filter(metad.pc, treat == "a"))
nrow(filter(metad.pc, treat == "p"))
```

### Scatterplot with GOI highlighted (Supplementary Figures 3 c)
* Treg signature genes highlighted
```{r, fig.height=2.6}
avp.pc.scat <- ggplot(avp.pc.tr.qstat.scat.filt) + 
  aes(x=logCPMpoly, y=logCPMallo) +
  scale_x_continuous(limits = c(-1,16),
                     breaks = c(0,3,6,9,12,15),
                     expand = c(0,0)
                     )+
  scale_y_continuous(limits = c(-1,16),
                     breaks = c(0,3,6,9,12,15),
                     expand = c(0,0)
                     )+
  geom_point(colour=rgb(50,50,50,30,maxColorValue = 255), shape=16, size = .375) +
  geom_point(data=filter(avp.pc.qstat.scat.filt.ann, goi.cat != "non DE"),
             aes(y=logCPMallo, x=logCPMpoly, color=goi.cat), 
             shape=16, 
             size = .5, 
             alpha=.75) +
  scale_color_manual(name = "Significant DEGs",
                     values = c("steelblue","orange"),
                     guide = guide_legend(override.aes = list(size = 3)))+
  geom_text_repel(data = avp.pc.qstat.scat.filt.ann.label,
                  aes(label=GeneSymbol, color = ann.col),
                  seed = 42,
                  direction = "both",
                  force = 500,
                  force_pull = 0,
                  size=1.66,
                  min.segment.length=0.0,
                  segment.size=0.15,
                  segment.color = avp.pc.qstat.scat.filt.ann.label$ann.col,
                  color = avp.pc.qstat.scat.filt.ann.label$ann.col,
                  max.overlaps = 30,
                  point.padding = 0.0,
                  fontface="italic")+
  ggtitle(label = "Signature genes in re-isolated Treg") +
  xlab(bquote(log[2]~CPM ~poly~Treg*" (n = 35)"))+
  ylab(bquote(log[2]~CPM ~allo~Treg*" (n = 36)"))+
  theme(plot.title = element_text(size = 10, face = "bold"),
        axis.title = element_text(size = 8),
        axis.text.x = element_text(size = 6, colour = "black"), #,angle = 90
        axis.text.y =element_text(size = 6, colour = "black"),
        axis.ticks.length.x = unit(0.1,"cm"),
        axis.ticks.length.y = unit(0.1,"cm"),
        axis.ticks = element_line(size=.236),
        plot.background = element_rect(fill="white",colour = "white"),
        panel.background = element_rect(fill = NA,color = "black", size=.236),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        legend.key = element_blank(),
        aspect.ratio = 1)
avp.pc.scat
```

* Herafter, the scatterplot is exported into the figure directory.

```{r}
pdf(file="./FIGDIR/SupplementaryFigure3c_TregDGEScatter_avp.pc_TregSignature.pdf", height=2.6, width=4.2)
avp.pc.scat
dev.off()
```

***

# GSEA with *in vitro* allo up Gene Set

## Reading in DGE from *in vitro* allo/poly comparison and GSEA
* GeneID lists of filtered and significantly up/down regulated genes in *in vitro* allo vs. poly Treg are generated in "Figure1_TregDGEScatter_GSEA_avp_invitro.Rmd".
* The results are read in and indices are generated for the current model of "allo versus poly in all BMT Treg".
* Using the fry() function, a rotation gene set testing for linear models is carried out.
```{r}
avp.ivt.tr.qstat.scat.filt.GeneID.UP <- read.delim(file = "./ANALDIR/Figure1e_TregDGE_avp_invitro.filt.sig.UP.txt")
avp.ivt.tr.qstat.scat.filt.GeneID.DN <- read.delim(file = "./ANALDIR/Figure1e_TregDGE_avp_invitro.filt.sig.DN.txt")
ind1 <- rownames(avp.pc.fit) %in% avp.ivt.tr.qstat.scat.filt.GeneID.UP[[1]]
ind2 <- rownames(avp.pc.fit) %in% avp.ivt.tr.qstat.scat.filt.GeneID.DN[[1]]
ind <- list(ind1,ind2)
names(ind) <- c("up in allo","up poly")
gse.avp.pc <- fry(dgel.avp.pc, index=ind, design=avp.pc.design.dge,contrast=avp.pc.con)
gse.avp.pc
```

## GSEA Barcode plot (Supplementary Figure 3 d)
* Barcode plot presenting logFC (allo vs. poly transplanted & re-isolated Treg) ranked Genes that are upregulated in allo *in vitro* expanded Treg.
```{r, fig.height=3.2, fig.width=6}
avp.pc.bc.avp.ivt %<a-% {
  par(mar=c(4.1,2.1,0.1,0.1))
  barcodeplot(avp.pc.tr$table$logFC, 
              index=ind1, 
              labels=c("up in poly","allo"), 
              xlab = bquote(log[2]*"FC"),
              col.bars = "orange")
  par(new=TRUE)
  plot.new( )
  plot.window( xlim=c(-5,5), ylim=c(-5,5) )
  text(-5,4.5,"Orange bars: Genes up in allo expanded donor Treg", adj = c(0,.5))
  text(c(-5,-3.75),c(3.4,3.5),
     c(expression(italic(P)["adj."]*"<"),
       scientific_10x(
         plyr::round_any(
           x = gse.avp.pc["up in allo",]$FDR,
           accuracy = 10^(ceiling(
             log10(gse.avp.pc["up in allo",]$FDR))-2),
           f = ceiling),
         digits = 1)), adj = c(0,.5))
}
avp.pc.bc.avp.ivt
```

* Barcode plot is also exported.

```{r}
pdf(file="./FIGDIR/SupplementaryFigure3d_avp.cp.Barcodeplot_GenesUpInIvtAllo.pdf", height=3.2, width=6)
avp.pc.bc.avp.ivt
dev.off()
```

## Summary of Condition/Organ-wise Comparisons
* First, make sure the "Figure3_GSEA_prc.pro_avp.c.l.s.indiv.Rmd" has been run.
* The file "SupplementaryFigure3b_avp_prc.pro_c.l.s_GSEA_ivt.txt" has to be generated first.
* The individual FDR values for the GSEA are visualized as horizontal bar plot.
```{r, fig.height=1.8, fig.width=2.6}
res_pvalues_ivt<-read.table(file = "./ANALDIR/SupplementaryFigure3d_avp_prc.pro_c.l.s_GSEA_ivt.txt", header = TRUE)
res_pvalues_ivt$logFDR <- log10(res_pvalues_ivt$pivt)
res_pvalues_ivt.bar <- ggplot(res_pvalues_ivt) + 
  aes(x=samples, y=-logFDR ) + coord_flip() + 
  scale_x_discrete(limits=rev) +
  scale_y_continuous(expand = c(0,0.1)) +
  geom_bar(stat="identity", fill="orange") +
  ggtitle(label = "GSEA P values", subtitle = "individual allo-poly comparisons") +
  geom_hline(yintercept=-log10(0.05),linetype='dashed',color='red') +
  ylab(bquote(-log[10]*" "*italic(P)[adj.]))+
  xlab(c("GvHD  no GvHD")) +
  theme(plot.title = element_text(size = 10, face = "bold"),
        plot.subtitle = element_text(size = 6, face = "bold"),
        axis.title = element_text(size = 8),
        axis.text.x = element_text(size = 6, colour = "black"), 
        axis.text.y =element_text(size = 6, colour = "black"),
        axis.ticks.length.x = unit(0.1,"cm"),
        axis.ticks.length.y = unit(0.1,"cm"),
        axis.ticks = element_line(size=.236),
        plot.background = element_rect(fill="white",colour = "white"),
        panel.background = element_rect(fill ="white",color = NA),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank())
res_pvalues_ivt.bar
```

* The summary bar plot is also exported into the figure directory.

```{r}
pdf(file= "./FIGDIR/SupplementaryFigure3d_avp.org.appl.wise.pvalues.bar_GenesUpInIvtAllo.pdf", height=1.8, width=2.6)
res_pvalues_ivt.bar
dev.off()
```

***

# GSEA with camera

## Pathway enrichment analysis for Hallmark gene sets

* Using camera, we look whether pathways from hallmark collections are enriched in our regulated genes.

```{r}
#hallmark
Mm.H <- readRDS(url("http://bioinf.wehi.edu.au/MSigDB/v7.1/Mm.h.all.v7.1.entrez.rds"))
H.ind.avp.pc <- ids2indices(Mm.H, id=dgel.avp.pc$genes$EntrezID)
gst.camera.avp.pc <- camera.DGEList(dgel.avp.pc,index=H.ind.avp.pc,
                                    design=avp.pc.design.dge,
                                    contrast=avp.pc.con,inter.gene.cor=0.05)
gst.camera.avp.pc
```

## GSEA Barcode Plot for Interferon Alpha Response Hallmark Gene Set (Supplementary Figure 3 d)
* The top Hallmark gene set found enriched is presented as barcode plot, hereafter.  
```{r, fig.height=3.2, fig.width=6}
#hallmark HALLMARK_INTERFERON_ALPHA_RESPONSE 
avp.pc.bc.ifnar %<a-% {
  par(mar=c(4.1,2.1,0.1,0.1))
  barcodeplot(avp.pc.tr$table$logFC,
              index=H.ind.avp.pc[["HALLMARK_INTERFERON_ALPHA_RESPONSE"]], 
              xlab = bquote(log[2]*"FC"),
              labels=c("up in poly","allo"), 
              col.bars = "steelblue")
  par(new=TRUE)
  plot.new( )
  plot.window( xlim=c(-5,5), ylim=c(-5,5) )
  text(5,4.5,"Steelblue bars: HALLMARK_INTERFERON_ALPHA_RESPONSE", adj = c(1,.5))
  text(c(3,5),c(3.4,3.5),
     c(expression(italic(P)["adj."]*"<"),
       scientific_10x(
         plyr::round_any(
           x = gst.camera.avp.pc["HALLMARK_INTERFERON_ALPHA_RESPONSE",]$FDR,
           accuracy = 10^(ceiling(
             log10(gst.camera.avp.pc["HALLMARK_INTERFERON_ALPHA_RESPONSE",]$FDR))-2),
           f = ceiling),
         digits = 1)), adj = c(1,.5))
}
avp.pc.bc.ifnar
```

* Barcode plot is exported, hereafter.

```{r}
pdf(file= "./FIGDIR/SupplementaryFigure3f_avp.pc.Barcodeplot_HALL_IFN.pdf", height=3.2, width=6)
avp.pc.bc.ifnar
dev.off()
```

## Summary of Condition/Organ-wise Comparisons
* First, make sure the "Figure3_GSEA_prc.pro_avp.c.l.s.indiv.Rmd" has been run.
* The file "SupplementaryFigure3b_avp_prc.pro_c.l.s_GSEA_ifn.txt" has to be generated first.
* The individual FDR values for the GSEA are visualized as horizontal bar plot.
```{r, fig.height=1.9, fig.width=2.6}
res_pvalues_ifn<-read.table(file = "./ANALDIR/SupplementaryFigure3f_avp_prc.pro_c.l.s_GSEA_ifn.txt", header = TRUE)
res_pvalues_ifn$logFDR <- log10(res_pvalues_ifn$pifn)
res_pvalues_ifn.bar <- ggplot(res_pvalues_ifn) + 
  aes(x=samples, y=-logFDR ) + coord_flip() + 
  scale_x_discrete(limits=rev) +
  scale_y_continuous(expand = c(0,0.1)) +
  geom_bar(stat="identity", fill="steelblue") +
  ggtitle(label = "GSEA P values", subtitle = "individual allo-poly comparisons") +
  geom_hline(yintercept=-log10(0.05),linetype='dashed',color='red') +
  ylab(bquote(-log[10]*" "*italic(P)[adj.]))+
  xlab(c("GvHD  no GvHD")) +
  theme(plot.title = element_text(size = 10, face = "bold"),
        plot.subtitle = element_text(size = 6, face = "bold"),
        axis.title = element_text(size = 8),
        axis.text.x = element_text(size = 6, colour = "black"), 
        axis.text.y =element_text(size = 6, colour = "black"),
        axis.ticks.length.x = unit(0.1,"cm"),
        axis.ticks.length.y = unit(0.1,"cm"),
        axis.ticks = element_line(size=.236),
        plot.background = element_rect(fill="white",colour = "white"),
        panel.background = element_rect(fill ="white",color = NA),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank())
res_pvalues_ifn.bar
```

* The summary adj. p value bar plot is exported, hereafter.

```{r}
pdf(file= "./FIGDIR/SupplementaryFigure3f_avp.org.appl.wise.pvalues.bar_HALL_IFN.pdf", height=1.9, width=2.6)
res_pvalues_ifn.bar
dev.off()
```

***

# SessionInfo
```{r, echo=FALSE, warning=FALSE}
sessionInfo(package = NULL)
writeLines(capture.output(sessionInfo(package = NULL)), "./Rscripts/Figure3_TregDGEScatter_GSEA_cvnc.pc_BMT_SessionInfo.txt")
```

***

# Bibliography