---
title: "(Supplementary) Figure4 related organ-wise Treg DGE Scatter & GSEA plots"
subtitle: "Organ wise DGE scatter plots comparing Treg from GvHD to noGvHD transplant models (Figure 4 a,b,c) are generated.\nRespective GSEA analyses on selected Hallmark and KEGG gene sets shown in Supplementary Figure 4 a,b,c are produced."
author: "David Dittmar"
date: "`r Sys.Date()`"
output:
  html_document:
    df_print: paged
    fig_caption: yes
    number_sections: no
    toc: yes
    toc_float: yes
    code_folding: hide
---

***

```{r setup, include=FALSE}
# Set global chunk options
knitr::opts_chunk$set(echo = TRUE, include = TRUE)
knitr::opts_knit$set(root.dir = normalizePath(".."))
# ...packages.bib file is also written here
knitr::write_bib(.packages(), "Figure4_TregDGEScatter_GSEA_pvc.cls_packages.bib")
```

```{r}
# Load required counts, metadata and annotaion RDS files
counts.m<-readRDS("./RCTDIR/counts.m.rds")
metad.m<-readRDS("./METADIR/metad.m.rds")
stid<-readRDS("./METADIR/stid.rds")
```

```{r, include=FALSE}
# Load required libraries
library(edgeR)
library(AnnotationDbi)
library(org.Mm.eg.db) #for mouse
library(tibble)
library(dplyr)
library(ggplot2)
library(ggrepel)
library(pryr)
library(ecoflux)
```
* First, required data objects (counts, metadata, short transcript ID) and R packages are loaded. 
* This is the starting point for most analyses conducted on bulk RNA-seq data for the associated manuscript.

***

# Data Preparation
* For the analysis, the data and metadata have to be prepared and properly annotated.

## Annotation Data Frame
* An annotation data frame with 4 columns is constructed.
* The additional EntrezID column is required for possible downstream pathway analyses or annotation operations.
```{r}
genes.df<-as.data.frame(strsplit2(stid,"[$]"))
colnames(genes.df)<-c("EnsemblID","GeneSymbol","Length","GeneType")
genes.df$EnsemblID<-as.character(genes.df$EnsemblID)
genes.df$GeneSymbol<-as.character(genes.df$GeneSymbol)
genes.df$Length<-as.numeric(as.character(genes.df$Length))
genes.df$GeneType<-as.character(genes.df$GeneType)
genes.df$EntrezID<- mapIds(org.Mm.eg.db,
                    keys=genes.df$GeneSymbol,
                    column="ENTREZID",
                    keytype="SYMBOL",
                    multiVals="first")
rownames(genes.df)<-stid
head(genes.df, n=5L)
#See if the gene length values are still as they should, so this should yield an empty vector
which(strsplit2(stid,"[$]")[,3] %in% genes.df$Length==FALSE)
```

## Subsetting Indices
* In this analysis, we focus on the organ wise comparison of Treg in transplant models to single out the Tconv effects: **BMT w Tconv vs. BMT w/o Tconv**.
* To that end, we require 3 models, one for each organ.
* Subsetting indices are generated accordingly.

### Colon
```{r}
cpc.ind<-(grepl("pro",metad.m$appl)|grepl("prc",metad.m$appl))&
  grepl("C", metad.m$org)
print(paste(sum(cpc.ind),"samples enter the analysis for Treg re-isolated from BMT colons!"))
```

### Liver
```{r}
lpc.ind<-(grepl("pro",metad.m$appl)|grepl("prc",metad.m$appl))&
  grepl("L", metad.m$org)
print(paste(sum(lpc.ind),"samples enter the analysis for Treg re-isolated from BMT livers!"))
```
### Spleen
```{r}
spc.ind<-(grepl("pro",metad.m$appl)|grepl("prc",metad.m$appl))&
  grepl("S", metad.m$org)
print(paste(sum(spc.ind),"samples enter the analysis for Treg re-isolated from BMT spleens!"))
```

## Counts
* Here, the comprehensive counts data frame is reduced using the created subsetting indices.

### Colon
* For colon:
```{r}
#Subsetting the respective counts object (counts.m)
counts.cpc<-counts.m[,cpc.ind]
paste(ncol(counts.cpc),"samples are selected:")
colnames(counts.cpc)
```

### Liver
* For liver:
```{r}
#Subsetting the respective counts object (counts.m)
counts.lpc<-counts.m[,lpc.ind]
paste(ncol(counts.lpc),"samples are selected:")
colnames(counts.lpc)
```

### Spleen
* For spleen:
```{r}
#Subsetting the respective counts object (counts.m)
counts.spc<-counts.m[,spc.ind]
paste(ncol(counts.spc),"samples are selected:")
colnames(counts.spc)
```

## Metadata
* The metadata has to be reduced accordingly. 
* The original order of samples is introduced for downstream re-ordering of the read count table.
* For each analyzed organ, two group variables are defined distinguishing either only Tconv presence in the model or additionally accounting for the expansion mehtod used for transplanted Treg.

### Colon
```{r}
metad.cpc<-metad.m[cpc.ind,]
metad.cpc$orig.ord<-1:nrow(metad.cpc)
metad.cpc<-metad.cpc[order(metad.cpc$appl),]
metad.cpc<-metad.cpc[order(metad.cpc$grp),]
counts.cpc.ind<-metad.cpc$orig.ord
head(metad.cpc, n = 4L)
#the group variable for all later assignments of colors/pchs.
group.cpc<-factor(as.character(metad.cpc$appl), levels = c("prc","pro"))
group.cpc
group.subcpc<-factor(as.character(metad.cpc$grp), levels = c("prc_C_a","prc_C_p","pro_C_a","pro_C_p"))
group.subcpc
```

### Liver
```{r}
metad.lpc<-metad.m[lpc.ind,]
metad.lpc$orig.ord<-1:nrow(metad.lpc)
metad.lpc<-metad.lpc[order(metad.lpc$appl),]
metad.lpc<-metad.lpc[order(metad.lpc$grp),]
counts.lpc.ind<-metad.lpc$orig.ord
head(metad.lpc, n = 4L)
#the group variable for all later assignments of colors/pchs.
group.lpc<-factor(as.character(metad.lpc$appl), levels = c("prc","pro"))
group.lpc
group.sublpc<-factor(as.character(metad.lpc$grp), levels = c("prc_L_a","prc_L_p","pro_L_a","pro_L_p"))
group.sublpc
```

### Spleen
```{r}
metad.spc<-metad.m[spc.ind,]
metad.spc$orig.ord<-1:nrow(metad.spc)
metad.spc<-metad.spc[order(metad.spc$appl),]
metad.spc<-metad.spc[order(metad.spc$grp),]
counts.spc.ind<-metad.spc$orig.ord
head(metad.spc, n = 4L)
#the group variable for all later assignments of colors/pchs.
group.spc<-factor(as.character(metad.spc$appl), levels = c("prc","pro"))
group.spc
group.subspc<-factor(as.character(metad.spc$grp), levels = c("prc_S_a","prc_S_p","pro_S_a","pro_S_p"))
group.subspc
```

## Re-ordering Counts Data Frame Columns
* Here, read counts are re-ordered according to the metadata objects.
* This is also performed for each organ individually.
* For colon:  
```{r}
counts.cpc.ord<-counts.cpc[,counts.cpc.ind]
head(counts.cpc.ord, n=4L)
counts.cpc<-counts.cpc.ord
rm(counts.cpc.ord)
```
* For Liver:  
```{r}
counts.lpc.ord<-counts.lpc[,counts.lpc.ind]
head(counts.lpc.ord, n=4L)
counts.lpc<-counts.lpc.ord
rm(counts.lpc.ord)
```
* And for Spleen:  
```{r}
counts.spc.ord<-counts.spc[,counts.spc.ind]
head(counts.spc.ord, n=4L)
counts.spc<-counts.spc.ord
rm(counts.spc.ord)
```

## Consistency Check
* In order to avoid wrongful labeling, sample_IDs are used to confirm that the counts and metadata objects are consistent.
```{r}
# Colon
if (sum(colnames(counts.cpc)==metad.cpc$sample_ID)==sum(cpc.ind)) {
  print("Colon Treg counts and metadata are properly subsetted and in accordance with each other. The analysis can be proceeded with!")
}else {
  print("Something is inconsistent, check the code above for errors during subsetting!")
}
# Liver
if (sum(colnames(counts.lpc)==metad.lpc$sample_ID)==sum(lpc.ind)) {
  print("Liver Treg counts and metadata are properly subsetted and in accordance with each other. The analysis can be proceeded with!")
}else {
  print("Something is inconsistent, check the code above for errors during subsetting!")
}
# Spleen
if (sum(colnames(counts.spc)==metad.spc$sample_ID)==sum(spc.ind)) {
  print("Spleen Treg counts and metadata are properly subsetted and in accordance with each other. The analysis can be proceeded with!")
}else {
  print("Something is inconsistent, check the code above for errors during subsetting!")
}
```

## DGE List Objects
* DGEList objects are generated for each organ, the main class of the edgeR package. 
* For full functionality, respective counts and group as well as the genes.df variables are used to do so.
```{r}
# For colon
dgel.cpc<-DGEList(counts = counts.cpc, group = group.subcpc, genes = genes.df)
# For liver
dgel.lpc<-DGEList(counts = counts.lpc, group = group.sublpc, genes = genes.df)
# For spleen
dgel.spc<-DGEList(counts = counts.spc, group = group.subspc, genes = genes.df)
head(dgel.cpc$samples, n=4L)
head(dgel.lpc$samples, n=4L)
head(dgel.spc$samples, n=4L)
```

## Exclusion of Rare Transcripts
* The "filterByExpr()" function is used to determine which genes have sufficiently large counts to be retained in a statistical analysis. 
* Only genes fulfilling that criterion are kept and also normalization factors are calculated.
```{r}
# For colon
keep.cpc <- filterByExpr(dgel.cpc)
dgel.cpc <- dgel.cpc[keep.cpc, , keep.lib.sizes=FALSE]
dgel.cpc <- calcNormFactors(dgel.cpc)
head(dgel.cpc$samples, n=4L)
print(paste(table(keep.cpc)[1],"genes are discarded while",table(keep.cpc)[2], "genes are retained in the colon-homing Treg analysis!"))
# For liver
keep.lpc <- filterByExpr(dgel.lpc)
dgel.lpc <- dgel.lpc[keep.lpc, , keep.lib.sizes=FALSE]
dgel.lpc <- calcNormFactors(dgel.lpc)
head(dgel.lpc$samples, n=4L)
print(paste(table(keep.lpc)[1],"genes are discarded while",table(keep.lpc)[2], "genes are retained in the liver-homing Treg analysis!"))
# For spleen
keep.spc <- filterByExpr(dgel.spc)
dgel.spc <- dgel.spc[keep.spc, , keep.lib.sizes=FALSE]
dgel.spc <- calcNormFactors(dgel.spc)
head(dgel.spc$samples, n=4L)
print(paste(table(keep.spc)[1],"genes are discarded while",table(keep.spc)[2], "genes are retained in the spleen-homing Treg analysis!"))
```

***

# Differential Gene Expression (DGE) analyses
* Differential Gene Expression (DGE) is performed in the following part.

## Designs for DGE
* Hereafter, designs are created that contain parameters that are of analytic interest as well as a blocking factor for the number of pcr2 cycles applied in library preparation.
* For each per-organ analysis, an independent design is created.
```{r}
# Colon
## Blocking Factor
pcr2.cpc<-factor(metad.cpc$PCR2.Cy)
treat.cpc<-factor(metad.cpc$treat)
## Design
cpc.design.dge<-model.matrix(~0+group.cpc+treat.cpc+pcr2.cpc)
colnames(cpc.design.dge)<-c("prc","pro","poly","twelve","thirteen","fourteen")
print("For colon-migrated Treg:")
as_tibble(cpc.design.dge)

# Liver
## Blocking Factor
pcr2.lpc<-factor(metad.lpc$PCR2.Cy)
treat.lpc<-factor(metad.lpc$treat)
## Design
lpc.design.dge<-model.matrix(~0+group.lpc+treat.lpc+pcr2.lpc)
colnames(lpc.design.dge)<-c("prc","pro","poly","thirteen","fourteen")
print("For liver-migrated Treg:")
as_tibble(lpc.design.dge)

# Spleen
## Blocking Factor
pcr2.spc<-factor(metad.spc$PCR2.Cy)
treat.spc<-factor(metad.spc$treat)
## Design
spc.design.dge<-model.matrix(~0+group.spc+treat.spc+pcr2.spc)
colnames(spc.design.dge)<-c("prc","pro","poly","thirteen","fourteen")
print("For spleen-migrated Treg:")
as_tibble(spc.design.dge)
```

## Dispersion Esitmation
* Hereafter, the dispersion is estimated. 
* This happens for each of the DGEList objects under consideration of the respective design defined above.
```{r}
# colon
dgel.cpc<-estimateDisp(dgel.cpc,design=cpc.design.dge,robust=TRUE)
print(paste("For the colon model, the common dispersion amounts to", round(dgel.cpc$common.dispersion,3)))
# liver
dgel.lpc<-estimateDisp(dgel.lpc,design=lpc.design.dge,robust=TRUE)
print(paste("For the liver model, the common dispersion amounts to", round(dgel.lpc$common.dispersion,3)))
# spleen
dgel.spc<-estimateDisp(dgel.spc,design=spc.design.dge,robust=TRUE)
print(paste("For the spleen model, the common dispersion amounts to", round(dgel.spc$common.dispersion,3)))
```

## GLM Fitting
* Fitting gene wise glms (generalized linear model) for the constructed designs.
```{r}
cpc.fit<-glmQLFit(dgel.cpc,cpc.design.dge)
lpc.fit<-glmQLFit(dgel.lpc,lpc.design.dge)
spc.fit<-glmQLFit(dgel.spc,spc.design.dge)
```

## DGE for Contrast of Interest (COI)
* The interest in this analysis lies in the differences between Treg re-isolated from mice that have received Tconv vs. no Tconv in addition to the Treg cell product.
* Again, this is laid out for each organ separately.
* Thus, our contrast of interest is pro - prc (**"prophylaxis - control BMT (w/o Tconv)"**) for the DGE analysis. 
* We can have a quick look at how many genes are DE with a FC > 1.5 and FDR < 0.05.  
```{r}
# contrasts
cpvc.con<-makeContrasts(pro - prc, levels=cpc.design.dge)
lpvc.con<-makeContrasts(pro - prc, levels=lpc.design.dge)
spvc.con<-makeContrasts(pro - prc, levels=spc.design.dge) 
# glmTreat results
cpvc.res <- glmTreat(cpc.fit, contrast=cpvc.con, lfc=log2(1.5))
lpvc.res <- glmTreat(lpc.fit, contrast=lpvc.con, lfc=log2(1.5))
spvc.res <- glmTreat(spc.fit, contrast=spvc.con, lfc=log2(1.5))
# qstat variable for storing results
cpvc.qstat<-topTags(cpvc.res,n=Inf)
lpvc.qstat<-topTags(lpvc.res,n=Inf)
spvc.qstat<-topTags(spvc.res,n=Inf)
# summaries
print("For colon:")
cpvc.qdt<-decideTestsDGE(cpvc.res)
summary(cpvc.qdt)
print("For liver:")
lpvc.qdt<-decideTestsDGE(lpvc.res)
summary(lpvc.qdt)
print("For spleen:")
spvc.qdt<-decideTestsDGE(spvc.res)
summary(spvc.qdt)
```


### Exporting DGE Statistics
* The data frames contained in the glmTreat function results is stored for direct access.
* This is done for each analysis (organ).
```{r}
# colon
write.table(x = cpvc.qstat$table,
            file = "./ANALDIR/Figure4c_TregDGE_GvHD.vs.noGvHD_colon_FC1.5.txt",
            quote=FALSE, 
            sep = "\t",
            col.names = TRUE,
            row.names = FALSE)
# liver
write.table(x = lpvc.qstat$table,
            file = "./ANALDIR/Figure4b_TregDGE_GvHD.vs.noGvHD_liver_FC1.5.txt",
            quote=FALSE, 
            sep = "\t",
            col.names = TRUE,
            row.names = FALSE)
# spleen
write.table(x = spvc.qstat$table,
            file = "./ANALDIR/Figure4a_TregDGE_GvHD.vs.noGvHD_spleen_FC1.5.txt",
            quote=FALSE, 
            sep = "\t",
            col.names = TRUE,
            row.names = FALSE)
```

***

# DGE Visualization
* Hereafter, DEGs are visualized *via* a scatter plot.
* Several sets of GOI (genes of interest) will be highlighted in different colors.

## Data Preparation
* The data is filtered based on logCPM and logRPKM thresholds.
* Necessary data transformations for the scatter plot are carried out as well: columns with individual logCPMs (calculated from logCPM and logFC)
```{r}
# colon
print("For colon:")
# Extraction of data frame from glmTreat result
cpvc.qstat.scat <- cpvc.qstat$table %>% #is a data frame
  # Introduction of logRPKM column
  mutate(logRPKM=logCPM-log2(Length * 0.001))
# Commonly used filter criteria
cpvc.qstat.scat.filt <- filter(cpvc.qstat.scat, 
                                      logCPM > 1 & logRPKM > 1) %>%
  # Transformation for scatter plot --> individual log CPMs
  mutate(logCPMprc = logCPM - logFC/2, # BMT control
         logCPMpro = logCPM + logFC/2) # GvHD
nrow(cpvc.qstat.scat.filt)
# Separation of significantly up and down regulated genes
cpvc.qstat.scat.filt.UP<-filter(cpvc.qstat.scat.filt, logFC >= 0 &
                                        FDR < 0.05)
cpvc.qstat.scat.filt.DN<-filter(cpvc.qstat.scat.filt, logFC <= 0 &
                                        FDR < 0.05)
nrow(cpvc.qstat.scat.filt.UP)
nrow(cpvc.qstat.scat.filt.DN)

# liver
print("For liver:")
# Extraction of data frame from glmTreat result
lpvc.qstat.scat <- lpvc.qstat$table %>% #is a data frame
  # Introduction of logRPKM column
  mutate(logRPKM=logCPM-log2(Length * 0.001))
# Commonly used filter criteria
lpvc.qstat.scat.filt <- filter(lpvc.qstat.scat, 
                                      logCPM > 1 & logRPKM > 1) %>%
  # Transformation for scatter plot --> individual log CPMs
  mutate(logCPMprc = logCPM - logFC/2, # BMT control
         logCPMpro = logCPM + logFC/2) # GvHD
nrow(lpvc.qstat.scat.filt)
# Separation of significantly up and down regulated genes
lpvc.qstat.scat.filt.UP<-filter(lpvc.qstat.scat.filt, logFC >= 0 &
                                        FDR < 0.05)
lpvc.qstat.scat.filt.DN<-filter(lpvc.qstat.scat.filt, logFC <= 0 &
                                        FDR < 0.05)
nrow(lpvc.qstat.scat.filt.UP)
nrow(lpvc.qstat.scat.filt.DN)

# spleen
print("For spleen:")
# Extraction of data frame from glmTreat result
spvc.qstat.scat <- spvc.qstat$table %>% #is a data frame
  # Introduction of logRPKM column
  mutate(logRPKM=logCPM-log2(Length * 0.001))
# Commonly used filter criteria
spvc.qstat.scat.filt <- filter(spvc.qstat.scat, 
                                      logCPM > 1 & logRPKM > 1) %>%
  # Transformation for scatter plot --> individual log CPMs
  mutate(logCPMprc = logCPM - logFC/2, # BMT control
         logCPMpro = logCPM + logFC/2) # GvHD
nrow(spvc.qstat.scat.filt)
# Separation of significantly up and down regulated genes
spvc.qstat.scat.filt.UP<-filter(spvc.qstat.scat.filt, logFC >= 0 &
                                        FDR < 0.05)
spvc.qstat.scat.filt.DN<-filter(spvc.qstat.scat.filt, logFC <= 0 &
                                        FDR < 0.05)
nrow(spvc.qstat.scat.filt.UP)
nrow(spvc.qstat.scat.filt.DN)
```

## Exporting UP and DN regulated Genes with background
* Significantly up and down regulated genes are exported in the following.
* For each analysis, you will find the respecitve gene lists in the analysis directory.
```{r}
#extract filtered gene lists with background genes
#colon
write.table(rownames(cpvc.qstat.scat.filt.UP),
            file = "./ANALDIR/Figure4c_TregDGE_GvHD.vs.noGvHD_colon.filt.sig.UP.txt",
            col.names = "GeneID",
            quote=FALSE, 
            row.names = FALSE)
write.table(rownames(cpvc.qstat.scat.filt.DN),
            file = "./ANALDIR/Figure4c_TregDGE_GvHD.vs.noGvHD_colon.filt.sig.DN.txt",
            col.names = "GeneID",
            quote=FALSE, 
            row.names = FALSE)
## Gene Symbols for Metascape
write.table(cpvc.qstat.scat.filt.UP$GeneSymbol,
            file = "./ANALDIR/Figure4c_TregDGE_GvHD.vs.noGvHD_colon.filt.sig.UP.genes.txt",
            col.names = "GeneSymbol",
            quote=FALSE, 
            row.names = FALSE)
write.table(cpvc.qstat.scat.filt.DN$GeneSymbol,
            file = "./ANALDIR/Figure4c_TregDGE_GvHD.vs.noGvHD_colon.filt.sig.DN.genes.txt",
            col.names = "GeneSymbol",
            quote=FALSE, 
            row.names = FALSE)
write.table(cpvc.qstat.scat.filt$GeneSymbol,
            file = "./ANALDIR/Figure4c_TregDGE_GvHD.vs.noGvHD_colon.filt.bg.genes.txt",
            col.names = "GeneSymbol",
            quote = FALSE,
            row.names = FALSE)
#liver
write.table(rownames(lpvc.qstat.scat.filt.UP),
            file = "./ANALDIR/Figure4b_TregDGE_GvHD.vs.noGvHD_liver.filt.sig.UP.txt",
            col.names = "GeneID",
            quote=FALSE, 
            row.names = FALSE)
write.table(rownames(lpvc.qstat.scat.filt.DN),
            file = "./ANALDIR/Figure4b_TregDGE_GvHD.vs.noGvHD_liver.filt.sig.DN.txt",
            col.names = "GeneID",
            quote=FALSE, 
            row.names = FALSE)
## Gene Symbols for Metascape
write.table(lpvc.qstat.scat.filt.UP$GeneSymbol,
            file = "./ANALDIR/Figure4b_TregDGE_GvHD.vs.noGvHD_liver.filt.sig.UP.genes.txt",
            col.names = "GeneSymbol",
            quote=FALSE, 
            row.names = FALSE)
write.table(lpvc.qstat.scat.filt.DN$GeneSymbol,
            file = "./ANALDIR/Figure4b_TregDGE_GvHD.vs.noGvHD_liver.filt.sig.DN.genes.txt",
            col.names = "GeneSymbol",
            quote=FALSE, 
            row.names = FALSE)
write.table(lpvc.qstat.scat.filt$GeneSymbol,
            file = "./ANALDIR/Figure4b_TregDGE_GvHD.vs.noGvHD_liver.filt.bg.genes.txt",
            col.names = "GeneSymbol",
            quote = FALSE,
            row.names = FALSE)
#spleen
write.table(rownames(spvc.qstat.scat.filt.UP),
            file = "./ANALDIR/Figure4a_TregDGE_GvHD.vs.noGvHD_spleen.filt.sig.UP.txt",
            col.names = "GeneID",
            quote=FALSE, 
            row.names = FALSE)
write.table(rownames(spvc.qstat.scat.filt.DN),
            file = "./ANALDIR/Figure4a_TregDGE_GvHD.vs.noGvHD_spleen.filt.sig.DN.txt",
            col.names = "GeneID",
            quote=FALSE, 
            row.names = FALSE)
## Gene Symbols for Metascape
write.table(spvc.qstat.scat.filt.UP$GeneSymbol,
            file = "./ANALDIR/Figure4a_TregDGE_GvHD.vs.noGvHD_spleen.filt.sig.UP.genes.txt",
            col.names = "GeneSymbol",
            quote=FALSE, 
            row.names = FALSE)
write.table(spvc.qstat.scat.filt.DN$GeneSymbol,
            file = "./ANALDIR/Figure4a_TregDGE_GvHD.vs.noGvHD_spleen.filt.sig.DN.genes.txt",
            col.names = "GeneSymbol",
            quote=FALSE, 
            row.names = FALSE)
write.table(spvc.qstat.scat.filt$GeneSymbol,
            file = "./ANALDIR/Figure4a_TregDGE_GvHD.vs.noGvHD_spleen.filt.bg.genes.txt",
            col.names = "GeneSymbol",
            quote = FALSE,
            row.names = FALSE)
```

## Scatterplots

### Definition of Genes of Interest (GOI)
* Hereafter, different sets of GOI are defined.
* Most importantly, genes related to suppression are highlighted for all three tissues to corroborate the protective role of colon Treg in the prophylaxis setting.  
```{r}
# spleen
## replication signature, purple Figure 4a
replication.sig <-c("Aurka","Smc2","Cdc20","Cdk1","E2f1")
## response to interferon beta/alpha, colored green in Figure 4a
ifn.res.spc.sig <- c("Ifit1", "Ifitm3","Iigp1","Lgals3bp",
                       "Ifi204","Irgm1" )
## R-MMU-5669034: TNFs bind their physiological receptors, colored pink in Figure 4a
tnf.bind.r.sig <-c("Lta","Tnfsf11","Tnfsf8","Tnfsf4","Tnfrsf25")

# liver
## response to interferon beta/alpha, colored green in Figure 4b
ifn.res.lpc.sig <- c("Ifit1", "Ifitm3","Iigp1","Lgals3bp",
                       "Ifi208","Irgm2")
## KEGG Pathway	mmu04659	Th17 cell differentiation --> upregulated in absence of Tconv-secreted Il2, colored blue in Figure 4b
th17.diff.sig <-c("Fos","Il6ra","Jun","Nfkbia","Rora")

#colon plot
## GO Biological Processes	GO:0046632	alpha-beta T cell differentiation, colored brown in Figure 4c
tcell.diff.sig <-c("Bcl6","Ccr7","Cd83","Rora","Tnfsf8","Tnfsf4")

#commonly highlighted suppression signature --> hand-picked, immunosuppression, other cannonical Treg functions and cytotoxicity, colored black in Figure 4a-c
supp.sig<-c("Gzmb","Il10","Gzma","Fasl","Prf1","Ctla4","Entpd1","Tgfb1","Tgfb3",
            "Il2ra","Nt5e","Fgl2","Ebi3","Cd274","Pdcd1lg2")
supp.sig <- sort(supp.sig)
```

### Color-coded geom_text_repel data frames

* Additional annotation data frames are constructed for easier plotting.

```{r}
#spleen
spvc.qstat.scat.filt.ann<-spvc.qstat.scat.filt%>%
  mutate(
    label.col = case_when(
      GeneSymbol %in% supp.sig ~ "black",
      GeneSymbol %in% replication.sig ~ "magenta3",
      GeneSymbol %in% ifn.res.spc.sig ~ "green4",
      GeneSymbol %in% tnf.bind.r.sig ~ "deeppink"
    ),
    goi.cat = case_when(
      logFC >= 0 & FDR < 0.05 ~ paste0("up in GvHD (",as.character(nrow(spvc.qstat.scat.filt.UP)),")"),
      logFC <= 0 & FDR < 0.05 ~ paste0("down in GvHD (",as.character(nrow(spvc.qstat.scat.filt.DN)),")"),
      TRUE ~ "non DE")
  )
spvc.qstat.scat.filt.ann.label<-spvc.qstat.scat.filt.ann[!is.na(spvc.qstat.scat.filt.ann$label.col),]

#liver
lpvc.qstat.scat.filt.ann<-lpvc.qstat.scat.filt%>%
  mutate(
    label.col = case_when(
      GeneSymbol %in% supp.sig ~ "black",
      GeneSymbol %in% ifn.res.lpc.sig ~ "green4",
      GeneSymbol %in% th17.diff.sig ~ "dodgerblue"
    ),
    goi.cat = case_when(
      logFC >= 0 & FDR < 0.05 ~ paste0("up in GvHD (",as.character(nrow(lpvc.qstat.scat.filt.UP)),")"),
      logFC <= 0 & FDR < 0.05 ~ paste0("down in GvHD (",as.character(nrow(lpvc.qstat.scat.filt.DN)),")"),
      TRUE ~ "non DE")
  )
lpvc.qstat.scat.filt.ann.label<-lpvc.qstat.scat.filt.ann[!is.na(lpvc.qstat.scat.filt.ann$label.col),]

#colon
cpvc.qstat.scat.filt.ann<-cpvc.qstat.scat.filt%>%
  mutate(
    label.col = case_when(
      GeneSymbol %in% supp.sig ~ "black",
      GeneSymbol %in% tcell.diff.sig ~ "sienna"
    ),
    goi.cat = case_when(
      logFC >= 0 & FDR < 0.05 ~ paste0("up in GvHD (",as.character(nrow(cpvc.qstat.scat.filt.UP)),")"),
      logFC <= 0 & FDR < 0.05 ~ paste0("down in GvHD (",as.character(nrow(cpvc.qstat.scat.filt.DN)),")"),
      TRUE ~ "non DE")
  )
cpvc.qstat.scat.filt.ann.label<-cpvc.qstat.scat.filt.ann[!is.na(cpvc.qstat.scat.filt.ann$label.col),]
```


### Spleen Scatterplot with GOI hightlighted (Figure 4a)
* Hereafter, the scatter plot for spleen samples is generated and plotted.
* black: supp.sig, purple: repl., green: ifn.I.response, pink: tnf.bind.r.sig
```{r, fig.height=2.6, fig.width=4.2, warning=FALSE}
length(which(metad.spc$appl =="prc"))
length(which(metad.spc$appl =="pro"))
spvc.scat <- ggplot(spvc.qstat.scat.filt) + 
  aes(x=logCPMprc, y=logCPMpro) +
  scale_x_continuous(limits = c(-1,16),
                     breaks = c(0,3,6,9,12,15),
                     expand = c(0,0)
                     )+
  scale_y_continuous(limits = c(-1,16),
                     breaks = c(0,3,6,9,12,15),
                     expand = c(0,0)
                     )+
  geom_point(colour=rgb(50,50,50,30,
                        maxColorValue = 255), 
             shape=16, 
             size = .375) +
  geom_point(data=filter(spvc.qstat.scat.filt.ann, goi.cat != "non DE"),
             aes(y=logCPMpro, x=logCPMprc, color=goi.cat), 
             shape=16,
             size = .5,
             alpha=.75) +
  scale_color_manual(name = "Significant DEGs",
                     values = c("deeppink","magenta"),
                     guide = guide_legend(override.aes = list(size = 3)))+
  geom_text_repel(data = spvc.qstat.scat.filt.ann.label,
                  aes(label=GeneSymbol, color = ann.col),
                  seed = 42,
                  direction = "both",
                  force = 500,
                  force_pull = 0,
                  size=1.66,
                  min.segment.length=0.0,
                  segment.size=0.15,
                  segment.color = spvc.qstat.scat.filt.ann.label$label.col,
                  color = spvc.qstat.scat.filt.ann.label$label.col,
                  max.overlaps = 30,
                  point.padding = 0.0,
                  fontface="italic")+
  ggtitle(label = "spleen, re-isolated Treg") +
  #xlab(bquote(log[2]~CPM ~poly~Treg*" (n = 35)"))+
  ylab(bquote(log[2]~CPM ~GvHD*" (n=12)"))+
  xlab(bquote(log[2]~CPM ~no~GvHD*" (n=12)"))+
  theme(plot.title = element_text(size = 10, face = "bold",colour = "deeppink4"),
        plot.subtitle = element_text(size = 6, face = "bold"),
        axis.title = element_text(size = 8),
        axis.text.x = element_text(size = 6, colour = "black"), #,angle = 90
        axis.text.y =element_text(size = 6, colour = "black"),
        axis.ticks.length.x = unit(0.1,"cm"),
        axis.ticks.length.y = unit(0.1,"cm"),
        axis.ticks = element_line(size=.236),
        plot.background = element_rect(fill="white",colour = "white"),
        panel.background = element_rect(fill = NA,color = "black", size=.236),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        legend.key = element_blank(),
        aspect.ratio = 1)
spvc.scat
```

* The scatterplot is also exported into the figure directory.  

```{r, warning=FALSE}
pdf(file="./FIGDIR/Figure4a_TregDGEScatter_GvHD.vs.noGvHD_spleen.SuppressionSignature.pdf", height=2.6, width=4.2)
spvc.scat
dev.off()
```

### Liver Scatterplot with GOI highlighted (Figure 4 b)
* Hereafter, the scatter plot for liver samples is generated and plotted.
* black: supp.sig, green: ifn.I.response, blue: th17.diff.sig
```{r, fig.height=2.6, fig.width=4.2, warning=FALSE}
length(which(metad.lpc$appl =="prc"))
length(which(metad.lpc$appl =="pro"))
lpvc.scat <- ggplot(lpvc.qstat.scat.filt) + 
  aes(x=logCPMprc, y=logCPMpro) +
  scale_x_continuous(limits = c(-1,16),
                     breaks = c(0,3,6,9,12,15),
                     expand = c(0,0)
                     )+
  scale_y_continuous(limits = c(-1,16),
                     breaks = c(0,3,6,9,12,15),
                     expand = c(0,0)
                     )+
  geom_point(colour=rgb(50,50,50,30,
                        maxColorValue = 255), 
             shape=16, 
             size = .375) +
  geom_point(data=filter(lpvc.qstat.scat.filt.ann, goi.cat != "non DE"),
             aes(y=logCPMpro, x=logCPMprc, color=goi.cat), 
             shape=16,
             size = .5,
             alpha=.75) +
  scale_color_manual(name = "Significant DEGs",
                     values = c("dodgerblue","steelblue"),
                     guide = guide_legend(override.aes = list(size = 3)))+
  geom_text_repel(data = lpvc.qstat.scat.filt.ann.label,
                  aes(label=GeneSymbol, color = ann.col),
                  seed = 42,
                  direction = "both",
                  force = 500,
                  force_pull = 0,
                  size=1.66,
                  min.segment.length=0.0,
                  segment.size=0.15,
                  segment.color = lpvc.qstat.scat.filt.ann.label$label.col,
                  color = lpvc.qstat.scat.filt.ann.label$label.col,
                  max.overlaps = 30,
                  point.padding = 0.0,
                  fontface="italic")+
  ggtitle(label = "liver, re-isolated Treg") +
  ylab(bquote(log[2]~CPM ~GvHD*" (n=11)"))+
  xlab(bquote(log[2]~CPM ~no~GvHD*" (n=12)"))+
  theme(plot.title = element_text(size = 10, face = "bold",colour = "steelblue4"),
        plot.subtitle = element_text(size = 6, face = "bold"),
        axis.title = element_text(size = 8),
        axis.text.x = element_text(size = 6, colour = "black"), #,angle = 90
        axis.text.y =element_text(size = 6, colour = "black"),
        axis.ticks.length.x = unit(0.1,"cm"),
        axis.ticks.length.y = unit(0.1,"cm"),
        axis.ticks = element_line(size=.236),
        plot.background = element_rect(fill="white",colour = "white"),
        panel.background = element_rect(fill = NA,color = "black", size=.236),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        legend.key = element_blank(),
        aspect.ratio = 1)
lpvc.scat
```

* The scatterplot is also exported into the figure directory.  

```{r, warning=FALSE}
pdf(file="./FIGDIR/Figure4b_TregDGEScatter_GvHD.vs.noGvHD_liver.SuppressionSignature.pdf", height=2.6, width=4.2)
lpvc.scat
dev.off()
```

### Colon Scatterplot with GOI highlighted (Figure 4 c)
* Hereafter, the scatter plot for colon samples is generated and plotted.
* black: supp.sig, brown: tcell.diff.sig
```{r, fig.height=2.6, fig.width=4.2, warning=FALSE}
length(which(metad.cpc$appl =="prc"))
length(which(metad.cpc$appl =="pro"))
cpvc.scat <- ggplot(cpvc.qstat.scat.filt) + 
  aes(x=logCPMprc, y=logCPMpro) +
  scale_x_continuous(limits = c(-1,16),
                     breaks = c(0,3,6,9,12,15),
                     expand = c(0,0)
                     )+
  scale_y_continuous(limits = c(-1,16),
                     breaks = c(0,3,6,9,12,15),
                     expand = c(0,0)
                     )+
  geom_point(colour=rgb(50,50,50,30,
                        maxColorValue = 255), 
             shape=16, 
             size = .375) +
  geom_point(data=filter(cpvc.qstat.scat.filt.ann, goi.cat != "non DE"),
             aes(y=logCPMpro, x=logCPMprc, color=goi.cat), 
             shape=16,
             size = .5,
             alpha=.75) +
  scale_color_manual(name = "Significant DEGs",
                     values = c("chocolate3","sienna"),
                     guide = guide_legend(override.aes = list(size = 3)))+
  geom_text_repel(data = cpvc.qstat.scat.filt.ann.label,
                  aes(label=GeneSymbol, color = ann.col),
                  seed = 42,
                  direction = "both",
                  force = 500,
                  force_pull = 0,
                  size=1.66,
                  min.segment.length=0.0,
                  segment.size=0.15,
                  segment.color = cpvc.qstat.scat.filt.ann.label$label.col,
                  color = cpvc.qstat.scat.filt.ann.label$label.col,
                  max.overlaps = 30,
                  point.padding = 0.0,
                  fontface="italic")+
  ggtitle(label = "colon, re-isolated Treg") +
  ylab(bquote(log[2]~CPM ~GvHD*" (n=12)"))+
  xlab(bquote(log[2]~CPM ~no~GvHD*" (n=12)"))+
  theme(plot.title = element_text(size = 10, face = "bold",colour = "sienna4"),
        plot.subtitle = element_text(size = 6, face = "bold"),
        axis.title = element_text(size = 8),
        axis.text.x = element_text(size = 6, colour = "black"), #,angle = 90
        axis.text.y =element_text(size = 6, colour = "black"),
        axis.ticks.length.x = unit(0.1,"cm"),
        axis.ticks.length.y = unit(0.1,"cm"),
        axis.ticks = element_line(size=.236),
        plot.background = element_rect(fill="white",colour = "white"),
        panel.background = element_rect(fill = NA,color = "black", size=.236),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        legend.key = element_blank(),
        aspect.ratio = 1)
cpvc.scat
```

* The scatterplot is also exported into the figure directory.  

```{r, warning=FALSE}
pdf(file="./FIGDIR/Figure4c_TregDGEScatter_GvHD.vs.noGvHD_colon.SuppressionSignature.pdf", height=2.6, width=4.2)
cpvc.scat
dev.off()
```

***

# GSEA with camera
* Following, gene set enrichment analyses (GSEA) for Hallmark and Kegg pathway associated genes are performed.
* DEG based on the comparison of prophylaxis vs. control BMT Treg are analyzed for colon, liver and spleen, separately.
* Using camera, we look whether pathways from Hallmark and Kegg gene sets are enriched in our regulated genes
* Barcode plots are used to visualize the results across the logFC ranking of genes based on the underlying comparison.  

## Pathway enrichment analysis for Kegg and Hallmark gene sets
* Hallmark gene sets are retrieved and gene identifiers for respective sets are converted into a list of indices.
* GSEA is performed using the "camera.DGEList" function on the DGEList object with the index and contrast specified previously.  
```{r}
# hallmark
Mm.H <- readRDS(
  url("http://bioinf.wehi.edu.au/MSigDB/v7.1/Mm.h.all.v7.1.entrez.rds"))
#Colon
H.ind.cpc <- ids2indices(Mm.H, id=dgel.cpc$genes$EntrezID)
# GSEA
gst.cpc.camera.hall <- camera.DGEList(
  dgel.cpc,index=H.ind.cpc,design=cpc.design.dge,
  contrast=cpvc.con,inter.gene.cor=0.05)
gst.cpc.camera.hall
# Liver
H.ind.lpc <- ids2indices(Mm.H, id=dgel.lpc$genes$EntrezID)
# GSEA
gst.lpc.camera.hall <- camera.DGEList(
  dgel.lpc,index=H.ind.lpc,design=lpc.design.dge,
  contrast=lpvc.con,inter.gene.cor=0.05)
gst.lpc.camera.hall
# Spleen
H.ind.spc <- ids2indices(Mm.H, id=dgel.spc$genes$EntrezID)
# GSEA
gst.spc.camera.hall <- camera.DGEList(
  dgel.spc,index=H.ind.spc,design=spc.design.dge,
  contrast=spvc.con,inter.gene.cor=0.05)
gst.spc.camera.hall
```
* Analogous retrieval of Kegg gene sets and index generation.
* GSEA on Kegg gene sets.  
```{r}
# Kegg
Mm.K <- readRDS(url("http://bioinf.wehi.edu.au/MSigDB/v7.1/Mm.c2.cp.kegg.v7.1.entrez.rds"))
# Colon
K.ind.cpc <- ids2indices(Mm.K, id=dgel.cpc$genes$EntrezID)
# GSEA
gst.cpc.camera.kegg <- camera.DGEList(dgel.cpc,index=K.ind.cpc,design=cpc.design.dge,contrast=cpvc.con,inter.gene.cor=0.05)
gst.cpc.camera.kegg
# Liver
K.ind.lpc <- ids2indices(Mm.K, id=dgel.lpc$genes$EntrezID)
# GSEA
gst.lpc.camera.kegg <- camera.DGEList(dgel.lpc,index=K.ind.lpc,design=lpc.design.dge,contrast=lpvc.con,inter.gene.cor=0.05)
gst.lpc.camera.kegg
# Spleen
K.ind.spc <- ids2indices(Mm.K, id=dgel.spc$genes$EntrezID)
# GSEA
gst.spc.camera.kegg <- camera.DGEList(dgel.spc,index=K.ind.spc,design=spc.design.dge,contrast=spvc.con,inter.gene.cor=0.05)
gst.spc.camera.kegg
```

## Barcode plot visualizations 
### HALLMARK_INTERFERON_ALPHA_RESPONSE (Supplementary Figure 4 a)
* Barcode plots representing the results for all three tissues.
* Colon, Liver and Spleen:  
```{r, fig.height=3.2, fig.width=6}
# Colon
cpc.bc.IFNA.pryr %<a-% {
  par(mar=c(4.1,2.1,0.1,0.1))
  barcodeplot(cpvc.res$table$logFC, 
            index=H.ind.cpc[["HALLMARK_INTERFERON_ALPHA_RESPONSE"]], 
            xlab = bquote(log[2]*"FC"),
            labels = c("up in Control BMT","GvHD colon"),
            col.bars = "sienna")
  par(new=TRUE)
  plot.new( )
  plot.window( xlim=c(-5,5), ylim=c(-5,5) )
  text(-5,4.8,"HALLMARK_INTERFERON_ALPHA_RESPONSE", adj = c(0,.5))
  text(c(-5,-3.75),c(3.4,3.5),
     c(expression(italic(P)["adj."]*"<"),
       scientific_10x(
         plyr::round_any(
           x = gst.cpc.camera.hall["HALLMARK_INTERFERON_ALPHA_RESPONSE",]$FDR,
           accuracy = 10^(ceiling(
             log10(gst.cpc.camera.hall["HALLMARK_INTERFERON_ALPHA_RESPONSE",]$FDR))-2),
           f = ceiling),
         digits = 1)), adj = c(0,.5))
}
cpc.bc.IFNA.pryr

# Liver
lpc.bc.IFNA.pryr %<a-% {
  par(mar=c(4.1,2.1,0.1,0.1))
  barcodeplot(lpvc.res$table$logFC, 
            index=H.ind.lpc[["HALLMARK_INTERFERON_ALPHA_RESPONSE"]], 
            xlab = bquote(log[2]*"FC"),
            labels = c("up in Control BMT","GvHD liver"),
            col.bars = "dodgerblue")
  par(new=TRUE)
  plot.new( )
  plot.window( xlim=c(-5,5), ylim=c(-5,5) )
  text(-5,4.8,"HALLMARK_INTERFERON_ALPHA_RESPONSE", adj = c(0,.5))
  text(c(-5,-3.75),c(3.4,3.5),
     c(expression(italic(P)["adj."]*"<"),
       scientific_10x(
         plyr::round_any(
           x = gst.lpc.camera.hall["HALLMARK_INTERFERON_ALPHA_RESPONSE",]$FDR,
           accuracy = 10^(ceiling(
             log10(gst.lpc.camera.hall["HALLMARK_INTERFERON_ALPHA_RESPONSE",]$FDR))-2),
           f = ceiling),
         digits = 1)), adj = c(0,.5))
}
lpc.bc.IFNA.pryr
# Spleen
spc.bc.IFNA.pryr %<a-% {
  par(mar=c(4.1,2.1,0.1,0.1))
  barcodeplot(spvc.res$table$logFC, 
            index=H.ind.spc[["HALLMARK_INTERFERON_ALPHA_RESPONSE"]], 
            xlab = bquote(log[2]*"FC"),
            labels = c("up in Control BMT","GvHD liver"),
            col.bars = "deeppink")
  par(new=TRUE)
  plot.new( )
  plot.window( xlim=c(-5,5), ylim=c(-5,5) )
  text(-5,4.8,"HALLMARK_INTERFERON_ALPHA_RESPONSE", adj = c(0,.5))
  text(c(-5,-3.75),c(3.4,3.5),
     c(expression(italic(P)["adj."]*"<"),
       scientific_10x(
         plyr::round_any(
           x = gst.spc.camera.hall["HALLMARK_INTERFERON_ALPHA_RESPONSE",]$FDR,
           accuracy = 10^(ceiling(
             log10(gst.spc.camera.hall["HALLMARK_INTERFERON_ALPHA_RESPONSE",]$FDR))-2),
           f = ceiling),
         digits = 1)), adj = c(0,.5))
}
spc.bc.IFNA.pryr
```
* All three barplots are exported into the figure directory, hereafter.
```{r}
pdf(file="./FIGDIR/SupplementaryFigure4a_Barcodeplot_pvc.colon_Hallmark_IFNA.pdf", height=3.2, width=6)
cpc.bc.IFNA.pryr
dev.off()
pdf(file="./FIGDIR/SupplementaryFigure4a_Barcodeplot_pvc.liver_Hallmark_IFNA.pdf", height=3.2, width=6)
lpc.bc.IFNA.pryr
dev.off()
pdf(file="./FIGDIR/SupplementaryFigure4a_Barcodeplot_pvc.spleen_Hallmark_IFNA.pdf", height=3.2, width=6)
spc.bc.IFNA.pryr
dev.off()
```

### HALLMARK_TNFA_SIGNALING_VIA_NFKB (Supplementary Figure 4 c)
* Barcode plots representing the results for all three tissues.
* Colon, Liver and Spleen:  
```{r, fig.height=3.2, fig.width=6}
# Colon
cpc.bc.TNFA.pryr %<a-% {
  par(mar=c(4.1,2.1,0.1,0.1))
  barcodeplot(cpvc.res$table$logFC, 
            index=H.ind.cpc[["HALLMARK_TNFA_SIGNALING_VIA_NFKB"]], 
            xlab = bquote(log[2]*"FC"),
            labels = c("up in Control BMT","GvHD colon"),
            col.bars = "sienna")
  par(new=TRUE)
  plot.new( )
  plot.window( xlim=c(-5,5), ylim=c(-5,5) )
  text(-5,4.8,"HALLMARK_TNFA_SIGNALING_VIA_NFKB", adj = c(0,.5))
  text(c(-5,-3.75),c(3.4,3.5),
     c(expression(italic(P)["adj."]*"<"),
       scientific_10x(
         plyr::round_any(
           x = gst.cpc.camera.hall["HALLMARK_TNFA_SIGNALING_VIA_NFKB",]$FDR,
           accuracy = 10^(ceiling(
             log10(gst.cpc.camera.hall["HALLMARK_TNFA_SIGNALING_VIA_NFKB",]$FDR))-2),
           f = ceiling),
         digits = 1)), adj = c(0,.5))
}
cpc.bc.TNFA.pryr

# Liver
lpc.bc.TNFA.pryr %<a-% {
  par(mar=c(4.1,2.1,0.1,0.1))
  barcodeplot(lpvc.res$table$logFC, 
            index=H.ind.lpc[["HALLMARK_TNFA_SIGNALING_VIA_NFKB"]], 
            xlab = bquote(log[2]*"FC"),
            labels = c("up in Control BMT","GvHD liver"),
            col.bars = "dodgerblue")
  par(new=TRUE)
  plot.new( )
  plot.window( xlim=c(-5,5), ylim=c(-5,5) )
  text(-5,4.8,"HALLMARK_TNFA_SIGNALING_VIA_NFKB", adj = c(0,.5))
  text(c(-5,-3.75),c(3.4,3.5),
     c(expression(italic(P)["adj."]*"<"),
       scientific_10x(
         plyr::round_any(
           x = gst.lpc.camera.hall["HALLMARK_TNFA_SIGNALING_VIA_NFKB",]$FDR,
           accuracy = 10^(ceiling(
             log10(gst.lpc.camera.hall["HALLMARK_TNFA_SIGNALING_VIA_NFKB",]$FDR))-2),
           f = ceiling),
         digits = 1)), adj = c(0,.5))
}
lpc.bc.TNFA.pryr
# Spleen
spc.bc.TNFA.pryr %<a-% {
  par(mar=c(4.1,2.1,0.1,0.1))
  barcodeplot(spvc.res$table$logFC, 
            index=H.ind.spc[["HALLMARK_TNFA_SIGNALING_VIA_NFKB"]], 
            xlab = bquote(log[2]*"FC"),
            labels = c("up in Control BMT","GvHD liver"),
            col.bars = "deeppink")
  par(new=TRUE)
  plot.new( )
  plot.window( xlim=c(-5,5), ylim=c(-5,5) )
  text(-5,4.8,"HALLMARK_TNFA_SIGNALING_VIA_NFKB", adj = c(0,.5))
  text(c(-5,-3.75),c(3.4,3.5),
     c(expression(italic(P)["adj."]*"<"),
       scientific_10x(
         plyr::round_any(
           x = gst.spc.camera.hall["HALLMARK_TNFA_SIGNALING_VIA_NFKB",]$FDR,
           accuracy = 10^(ceiling(
             log10(gst.spc.camera.hall["HALLMARK_TNFA_SIGNALING_VIA_NFKB",]$FDR))-2),
           f = ceiling),
         digits = 1)), adj = c(0,.5))
}
spc.bc.TNFA.pryr
```
* All three barplots are exported into the figure directory, hereafter.
```{r}
pdf(file="./FIGDIR/SupplementaryFigure4c_Barcodeplot_pvc.colon_Hallmark_TNFA.pdf", height=3.2, width=6)
cpc.bc.TNFA.pryr
dev.off()
pdf(file="./FIGDIR/SupplementaryFigure4c_Barcodeplot_pvc.liver_Hallmark_TNFA.pdf", height=3.2, width=6)
lpc.bc.TNFA.pryr
pdf(file="./FIGDIR/SupplementaryFigure4c_Barcodeplot_pvc.spleen_Hallmark_TNFA.pdf", height=3.2, width=6)
spc.bc.TNFA.pryr
dev.off()
```

### KEGG_DNA_REPLICATION (Supplementary Figure 4 b)
* Barcode plots representing the results for all three tissues.
* Colon, Liver and Spleen:  
```{r, fig.height=3.2, fig.width=6}
# Colon
cpc.bc.KREP.pryr %<a-% {
  par(mar=c(4.1,2.1,0.1,0.1))
  barcodeplot(cpvc.res$table$logFC, 
            index=K.ind.cpc[["KEGG_DNA_REPLICATION"]], 
            xlab = bquote(log[2]*"FC"),
            labels = c("up in Control BMT","GvHD colon"),
            col.bars = "sienna")
  par(new=TRUE)
  plot.new( )
  plot.window( xlim=c(-5,5), ylim=c(-5,5) )
  text(-5,4.8,"KEGG_DNA_REPLICATION", adj = c(0,.5))
  text(c(-5,-3.75),c(3.4,3.5),
     c(expression(italic(P)["adj."]*"<"),
       scientific_10x(
         plyr::round_any(
           x = gst.cpc.camera.kegg["KEGG_DNA_REPLICATION",]$FDR,
           accuracy = 10^(ceiling(
             log10(gst.cpc.camera.kegg["KEGG_DNA_REPLICATION",]$FDR))-2),
           f = ceiling),
         digits = 1)), adj = c(0,.5))
}
cpc.bc.KREP.pryr

# Liver
lpc.bc.KREP.pryr %<a-% {
  par(mar=c(4.1,2.1,0.1,0.1))
  barcodeplot(lpvc.res$table$logFC, 
            index=K.ind.lpc[["KEGG_DNA_REPLICATION"]], 
            xlab = bquote(log[2]*"FC"),
            labels = c("up in Control BMT","GvHD liver"),
            col.bars = "dodgerblue")
  par(new=TRUE)
  plot.new( )
  plot.window( xlim=c(-5,5), ylim=c(-5,5) )
  text(-5,4.8,"KEGG_DNA_REPLICATION", adj = c(0,.5))
  text(c(-5,-3.75),c(3.4,3.5),
     c(expression(italic(P)["adj."]*"<"),
       scientific_10x(
         plyr::round_any(
           x = gst.lpc.camera.kegg["KEGG_DNA_REPLICATION",]$FDR,
           accuracy = 10^(ceiling(
             log10(gst.lpc.camera.kegg["KEGG_DNA_REPLICATION",]$FDR))-2),
           f = ceiling),
         digits = 1)), adj = c(0,.5))
}
lpc.bc.KREP.pryr
# Spleen
spc.bc.KREP.pryr %<a-% {
  par(mar=c(4.1,2.1,0.1,0.1))
  barcodeplot(spvc.res$table$logFC, 
            index=K.ind.spc[["KEGG_DNA_REPLICATION"]], 
            xlab = bquote(log[2]*"FC"),
            labels = c("up in Control BMT","GvHD liver"),
            col.bars = "deeppink")
  par(new=TRUE)
  plot.new( )
  plot.window( xlim=c(-5,5), ylim=c(-5,5) )
  text(-5,4.8,"KEGG_DNA_REPLICATION", adj = c(0,.5))
  text(c(-5,-3.75),c(3.4,3.5),
     c(expression(italic(P)["adj."]*"<"),
       scientific_10x(
         plyr::round_any(
           x = gst.spc.camera.kegg["KEGG_DNA_REPLICATION",]$FDR,
           accuracy = 10^(ceiling(
             log10(gst.spc.camera.kegg["KEGG_DNA_REPLICATION",]$FDR))-2),
           f = ceiling),
         digits = 1)), adj = c(0,.5))
}
spc.bc.KREP.pryr
```
* All three barplots are exported into the figure directory, hereafter.
```{r}
pdf(file="./FIGDIR/SupplementaryFigure4b_Barcodeplot_pvc.colon_Hallmark_KeggREP.pdf", height=3.2, width=6)
cpc.bc.KREP.pryr
dev.off()
pdf(file="./FIGDIR/SupplementaryFigure4b_Barcodeplot_pvc.liver_Hallmark_KeggREP.pdf", height=3.2, width=6)
lpc.bc.KREP.pryr
pdf(file="./FIGDIR/SupplementaryFigure4b_Barcodeplot_pvc.spleen_Hallmark_KeggREP.pdf", height=3.2, width=6)
spc.bc.KREP.pryr
dev.off()
```

***

# SessionInfo
```{r, echo=FALSE, warning=FALSE}
sessionInfo(package = NULL)
writeLines(capture.output(sessionInfo(package = NULL)), "./Rscripts/Figure4_TregDGEScatter_GSEA_pvc.cls_SessionInfo.txt")
```

***

# Bibliography